pkg/init/init.go

```diff
@@ -273,17 +274,27 @@ func (o InitClient) InitDevfile(flags map[string]string, contextDir string, preI
 		return nil
 	}
 
-	preInitHandlerFunc(len(flags) == 0)
+	if preInitHandlerFunc != nil {
+		preInitHandlerFunc(len(flags) == 0)
+	}
 
 	devfileObj, _, err := o.SelectAndPersonalizeDevfile(map[string]string{}, contextDir)
 	if err != nil {
 		return err
 	}
 
-	// Set the name in the devfile and writes the devfile back to the disk
-	err = o.PersonalizeName(devfileObj, map[string]string{})
+	// Set the name in the devfile but do not write it yet.
+	name, err := o.PersonalizeName(devfileObj, map[string]string{})
 	if err != nil {
 		return fmt.Errorf("failed to update the devfile's name: %w", err)
 	}
-	return nil
+	metadata := devfileObj.Data.GetMetadata()
+	metadata.Name = name
+	devfileObj.Data.SetMetadata(metadata)
+
+	if postInitHandlerFunc != nil {
+		err = postInitHandlerFunc(devfileObj)
+	}
+
```

Comment:
I tried returning `devfileObj` before, but this increased code duplication, as reported by Sonar Cloud checks, because the handling was exactly the same in both `dev.go` and `deploy.go` callers.
But maybe I should not worry that much about that issue for now :)

Also, I think I had another issue with returning a `parser.DevfileObj`, because there is a check before in this method that is expected to do nothing if there is already a Devfile file present:
https://github.com/redhat-developer/odo/blob/dbfb0a10363e6894072fdccc72bcee418d7592f1/pkg/init/init.go#L269-L276
In such case, I was reluctant to return an additional boolean indicating that a Devfile already exists (looks like this method is doing too many things to me).
Delegating this simple check to the callers would also result in too much code duplication as well.

Do you think there is a place in the codebase that I could leverage to avoid too much duplication?

Reply:
Good point. And with a little hindsight and from this point of view, your solution looks nice: we call the handler/write the file only when there is a devfileObj to be saved (you could rename the last parameter to reflect this)

/lgtm

Reply:
Done - [This new commit](https://github.com/redhat-developer/odo/pull/5553/commits/2462c97af749f0a0f44e8051f10a680930c74af0) should hopefully make the intent of the callback and its parameter clearer.